### PR TITLE
Deep Learning is here

### PR DESCRIPTION
**Pull request by bvelitchkine**
Ce que j'ai fait dans cette PR.

# Refactoring

J'ai repris ce que tu as fait, d'abord pour comprendre, et dans certains cas, pour modifier deux ou trois trucs, notamment cette histoire de *portfolio_cash* qui faussait totalement le calcul de ```get_current_returns```.

# Deep Reinforcement Learning
## Model

J'ai créé le *DenseModel* qu'on utilise pour estimer la q-value. C'est un modèle très simple, j'ai **scrupuleusement** repris celui du répo qu'on a plus ou moins pompé, en faisant simplement l'adaptation de keras vers tensorflow.

## Agent

J'ai ensuite créé le *DQNAgent*, l'agent basé sur le modèle précédemment créé. Il suit une **politique epsilon-greedy** : avec proba epsilon, il choisit une action aléatoire, avec probabilité 1-epsilon, l'action qui maximise la q-value.

**Je me suis inspiré du code du dernier cours de RL** du prof des Ponts. Son code est simple et bien organisé.

## Runners

Si les agents sont les marionnettes, les "runners" sont les marionnettistes. J'en ai implémenté de nouveaux : d'une part pour **entrainer un agent sur plusieurs épisodes**, d'autre part, pour **tester un agent pré-entrainé** sans le modifier.

# Pour la suite

Voici ce que je pense qu'il nous reste à faire (idéalement) : 

- Faire des fonctions d'affichage pour mieux visualiser ce qu'il se passe au cours des épisodes (e.g [valeur du portfolio + évolution du cours ETH/EUR] au fil du temps).
- Entrainer un agent sur l'année 2021 par exemple, et voir comment il se comporte sur l'année 2022.
- Revoir ce qu'on avait dit qu'on ferait dans la project proposal 🙃 
- Faire un README bien clean !

### CODE REVIEW

**Comment by leagonnet33** [comment]
Merci beaucoup pour tout le taf ! Ca me paraît niquel, j'ai juste quelques questions/remarques, c'est surtout sur la logique du code: 
- Pour la gestion de la quantité d'argent dans le portefeuille (et le fait qu'on ne peut pas acheter si on n'a pas le cash nécessaire), de ce que j'ai compris la logique actuelle c'est: à chaque étape, l'agent prend une décision acheter/vendre/ne rien faire sans contrainte particulière concernant le cash dispo. Ensuite, on essaye d'appliquer cette action au portefeuille, si le contenu du portefeuille fait qu'on ne peut pas effectuer l'action sélectionnée, en fait on ne fait rien, c'est bien ça ? Mais du coup, l'action que l'agent va garder en mémoire comme s'il l'avait effectuée est celle qu'il a choisie au départ, cela ne pose pas de problème ? Je sais pas si c'est clair ahah
- le reward n'est pas pris en compte dans la fonction `agent_step`, ça me paraît bizarre comme ça mais il y a peut être une subtilité que je n'ai pas saisi, tu pourrais me détailler un peu pourquoi ?
- Enfin sur l'aspect Deep Learning, comme je t'ai dit j'en ai jamais fait donc je t'avoue que je ne pense pas avoir été capable de repérer une erreur/imprécision dans ton code s'il y en avait une, mais si ça te va, je te fais confiance !
- Enfin pour les next steps je suis alignée avec toi ! Je suis repassée rapidement sur le project proposal et ce qu'on avait dit qu'on ferait et qui me semble envisageable là, c'est principalement: étudier l'influence de la taille de la `lookback_window`, l'influence des features décrivant chaque state, éventuellement tester un autre reward pour voir si les résultats bougent (si on a le temps/l'envie). Enfin on avait parlé de tester si les apprentissages d'un agent étaient "transitifs" d'une monnaie à l'autre (est-ce qu'un agent entraîné sur ETH peut être efficace sur BTC) donc je pense qu'on peut traiter cette question en dernier lieu si on s'ennuie

**Comment by bvelitchkine** [comment]
Point par point.

1. Oui c'est très clair. Ecoute c'est une bonne question, mais j'aurais tendance à dire que l'agent se rappellera que : dans telle situation, s'il prend telle décision (peu importe ce qu'il se passe en réalité sous le manteau), il reçoit telle ou telle récompense, et c'est la seule chose qui compte. Donc j'aurais tendance à dire que ça ne pose pas problème.

2. Comme tu peux le voir dans `train_dqn_agent`, à chaque étape, après chaque action prise et appliquée par l'agent, la reward est stockée en mémoire (dans _memory_). On utilise ensuite les valeurs stockées en mémoire dans la méthode _train_ du `DQNAgent`. Bref, on en tient compte, mais seulement au bon moment. En fait, pourquoi est-ce qu'on fait ça, et pourquoi est-ce qu'on ne train pas après chaque action prise, c'est pour des raisons de stabilité. Le prof des Ponts l'évoque et le montre en pratique dans le dernier cours d'ailleurs.

3. Bien aimable 🙃 aha. En réalité le modèle est simple, c'est une série de couches linéaires, donc sache que c'est facilement abordable si t'as le temps de regarder. Mais t'as jamais jamais fait de DL depuis ton arrivée à CS ?

4. Ok, avant tout ce que tu as évoqué, je pense qu'il nous faut vraiment des fonctions pour afficher les résultats, et mieux visualiser ce qu'il se passe pendant l'entrainement et après. Après ça, si ok pour toi, on peut se dire qu'on fait les choses dans cet ordre :
    - Influence de la _lookback_window_
    - Influence des métriques dont on tient compte dans le "state"
    - Entrainer un agent sur ETH, et voir ce qu'il fait sur BTC (c'est simple normalement), il y aura juste à charger le modèle pré-entrainé et à le faire tourner sur un jeu de données différent.
    > Gardons en tête que ce qu'on fait à chaque fois, c'est entrainer, puis tester. Il faudra ensuite valider en regardant le comportement de l'agent sur des données ETH/USDC qu'il n'a jamais vue (donc typiquement années 2022, ou 2020). Il faudra le faire à chaque fois qu'on changera la taille de la _lookback_window_ par exemple.

**Comment by leagonnet33** [comment]
@Bassvelitchkine merci beaucoup pour tes retours ! 

1. en effet tu as raison ça fait suffisamment sens comme ça!
2. ça marche c'est super clair, merci !
3. Je vais essayer de me pencher dessus en détail alors ! Non j'ai jamais fait de DL, tous les cours de DL cette année étaient pas compatibles avec l'alternance, ça tombait sur mes jours de taf... donc grosses lacunes sur ce sujet ahah
4. Yes je suis totalement dacc avec toi que la priorité est d'avoir des fonctions de visualisation et ensuite on pourra éventuellement aller plus loin et jouer avec les paramètres. Je suis alignée avec l'ordre que tu proposes pour ça

Du coup, je merge!

**Review by leagonnet33 — APPROVED** [review]
LGTM
